### ISSUE TITLE
L_unc loss function collapse

### ISSUE THREAD

**Issue by yaoyuan10475** [ENSTA-U2IS/LDU#10]
Hello, I'm trying to reproduce a semantic segmentation task. The inputs to the L_unc loss function are the confidence score and the result of the task loss function (Cross Entropy with reduction='none'), which have the same shape [B, H, W]. Is that so?
But during the training process, I found that the L_unc loss function value dropped to -3429.994, and the segmentation results also became collapsed. What is the reason for this?
I would be very grateful if I could get your reply.

**Comment by giannifranchi** [comment]
You have to put some weights among the different losses. Your question is not clear could you share your code and I will see how I can answer it. No guarantee I can answer. To be honest the task is quite simple you just need to learn to predict the error, that is the uncertainty loss. Multiple papers do the same.

**Comment by yaoyuan10475** [comment]
Absolutely okay. It's my honor to get your advice. I have solved the problem just now. It is because in the L_unc loss function, the result of the L_task loss function does not have detach(). Before modification:
```
loss_uncer = criterion_uncer(uncer.squeeze(dim=1), loss_task)
```
after modification：
```
loss_uncer = criterion_uncer(uncer.squeeze(dim=1), loss_task.detach())
```
Although the problem just now has been solved, I am not sure whether the L_unc loss function is correct. Could you take a look at it for me? 
There is another question, is ```nb Prototypes``` the same number of target classes as semantic segmentation?
The complete training code is as follows:
```
import torch
from model import build_model
from mmengine import Config
from torch.serialization import FILE_LIKE
from torch import nn
import os
import time
import datetime
import torch.backends.cudnn as cudnn
import numpy as np
import random
from utils.optimizer import build_optimizer
from utils.lr_scheduler import build_lr_scheduler
from utils.metric import seg_evaluate
from tensorboardX import SummaryWriter
import torch.nn.functional as F
from dataset import build_dataloader
from torch.cuda.amp import autocast, GradScaler



def setup_seed(seed):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    cudnn.deterministic = True
    cudnn.benchmark = False
    cudnn.enabled = False


class entropy_loss(nn.Module):
    def __init__(self):
        super(entropy_loss, self).__init__()

    def forward(self, embedding):
        embedding = nn.Softmax(dim=1)(embedding)
        minus_entropy = embedding * torch.log(embedding)
        minus_entropy = torch.sum(minus_entropy, dim=1)
        return minus_entropy.mean()


class uncertainty_loss(nn.Module):
    def __init__(self):
        super(uncertainty_loss, self).__init__()


    def forward(self, uncer, task_loss):
        loss = nn.BCEWithLogitsLoss(pos_weight = torch.tensor([5.0]).to(uncer.device), reduction='mean')(uncer, task_loss)
        return loss


class dissimilar_loss(nn.Module):
    def __init__(self):
        super(dissimilar_loss, self).__init__()

    def forward(self, protos):
        loss = -1 * torch.mean(torch.cdist(protos, protos))
        return loss


def main(cfg):

    device  = cfg.train.device
    # dataloader
    train_loader, val_loader = build_dataloader(cfg)
    
    # model creating
    model = build_model(cfg)

    # optimizer
    optimizer = build_optimizer(cfg, model)
    scaler = GradScaler()
    # lr scheduler
    lr_scheduler = build_lr_scheduler(cfg, optimizer, len(train_loader))
    # loss
    criterion_uncer = uncertainty_loss()
    criterion_entro = entropy_loss()
    criterion_dissi = dissimilar_loss()
    
    # eval
    evaluation = seg_evaluate(num_classes = 2, ignore_index=-1)
    # resume
    if cfg.train.resume:
        checkpoint = torch.load(cfg.train.model_checkpoint, map_location='cpu')
        start_epoch = checkpoint['epoch'] + 1
        model.load_state_dict(checkpoint['model'])
        optimizer.load_state_dict(checkpoint['optimizer'])
        lr_scheduler.load_state_dict(checkpoint['lr_scheduler'])
        del checkpoint
    else:
        start_epoch=0
    model.to(device)
    # train settings
    best_miou = 0.
    loss_value = 0.
    epochs = cfg.train.epochs
    train_loader_length = len(train_loader)
    val_loader_length = len(val_loader)
    total_iter = epochs * train_loader_length
    inter_log = cfg.train.inter_log
    
    # 分布式训练
    model.train()
  
    # workdir to save model.pth and writer    

    work_dir = os.path.join('work_dir', cfg.train.work_dir)
    if not os.path.exists(work_dir):
        os.makedirs(work_dir)
    writer = SummaryWriter(work_dir)

    
    start_time = time.time()
    for epoch in range(start_epoch, epochs):
        for i, (batch) in enumerate(train_loader, start=1):
            iter = epoch * train_loader_length + i
            img, mask = batch['img'].to(device), batch['mask'].to(device)
            weight_loss = batch["weight_loss"].to(device)
            with autocast():
                # final_depth final_uncer omega embedding_
                seg_logits, uncer, omega, embedding_ = model(img) #seg_logits: [B, C, H, W], uncer: [B, 1, H, W]
                seg_logits = F.interpolate(input=seg_logits, size=mask.shape[-2:], mode='bilinear', align_corners=False)
                uncer = F.interpolate(input=uncer, size=mask.shape[-2:], mode='bilinear', align_corners=False)
            
                loss_task = F.cross_entropy(seg_logits, mask, reduction='none')
                loss_uncer = criterion_uncer(uncer.squeeze(dim=1), loss_task.detach())
                loss_task = torch.mean(loss_task * weight_loss)

                loss_omega = criterion_entro(embedding_)
                loss_dissi = criterion_dissi(omega)
                
                loss_target = loss_task + (loss_uncer + loss_omega + loss_dissi) * 0.1
     
            optimizer.zero_grad()

            scaler.scale(loss_target).backward()
            lr_scheduler.step_update(iter)
   
            scaler.step(optimizer)
            scaler.update()

            loss_value += loss_target.item()
            # eval
            evaluation.update(np.asarray(seg_logits.argmax(1).detach().cpu()), np.asarray(mask.detach().cpu()))
          
            if iter%inter_log==0:
                seg_iou, _, _, _ = evaluation.get_scores()
                evaluation.reset()
                lr = optimizer.param_groups[0]['lr']
                # 计算训练时间
                eta_seconds = ((time.time() - start_time) / iter) * (total_iter - iter)
                eta_string = str(datetime.timedelta(seconds=int(eta_seconds)))
                print(f"{iter}/{total_iter}--Cost Time: {str(datetime.timedelta(seconds=int(time.time() - start_time)))} || Estimated Time: {eta_string} || lr: {lr} ||\
                      seg_iou:{seg_iou} || loss value:{loss_value/inter_log} ||\
                        loss uncer: {loss_uncer.item()} || loss omega: {loss_omega.item()} || loss dissi: {loss_dissi.item()}")
 
                writer.add_scalar('loss all', loss_value/inter_log, iter)
                writer.add_scalar('lr', lr, iter)
                writer.add_scalar('seg_iou', seg_iou, iter)

                writer.add_scalar('loss task', loss_task.item(), iter)
                writer.add_scalar('loss uncer', loss_uncer.item(), iter)
                writer.add_scalar('loss omega', loss_omega.item(), iter)
                writer.add_scalar('loss dissi', loss_dissi.item(), iter)
                
                loss_value = 0.
        
        # validation
        model.eval()
        loss_value = 0.
        evaluation.reset()
        for val_iter, batch in enumerate(val_loader):
            img, mask = batch['img'].to(device), batch['mask'].to(device)
            
            with autocast():
                with torch.no_grad():
                    seg_logits, _ = model(img)

            seg_logits = F.interpolate(input=seg_logits, size=mask.shape[-2:], mode='bilinear', align_corners=False)

            evaluation.update(np.asarray(seg_logits.argmax(1).detach().cpu()), np.asarray(mask.detach().cpu()))
        # 计算评价指标
        mean_IoU, (neg_iou, pos_iou), _, _ = evaluation.get_scores()
        print(f'pos_iou: {pos_iou}, neg_iou: {neg_iou}, mIoU: {mean_IoU}')
        evaluation.reset()
        writer.add_scalar('val miou', mean_IoU, iter)
        writer.add_scalar('val pos_iou', pos_iou, iter)
        writer.add_scalar('val neg_iou', neg_iou, iter)
        torch.save({'model':model.state_dict(), 'optimizer':optimizer.state_dict(),
                    'lr_scheduler':lr_scheduler.state_dict(), 'epoch':epoch, 'scaler': scaler.state_dict()},
                    os.path.join(work_dir, f'{cfg.train.work_dir}_last.pth'))
        if mean_IoU > best_miou:
            best_miou = mean_IoU
            torch.save({'model':model.state_dict(), 'optimizer':optimizer.state_dict(),
                    'lr_scheduler':lr_scheduler.state_dict(), 'epoch':epoch, 'scaler': scaler.state_dict()},
                    os.path.join(work_dir, "{}_best_{:.4f}.pth".format(cfg.train.work_dir, best_miou*100)))
        model.train()

if __name__ == "__main__":
    setup_seed(42)
    cfg = Config.fromfile('config/hrnet/ldu_spd_lite_hrnet18_neck.py')
    main(cfg)


    
```

**Comment by xuanlongORZ** [comment]
I roughly went through the code (mainly the uncertainty loss function part). In 'class uncertainty_loss', you didn't normalize the task loss (which is mentioned on the line after the definition of $L^{Unc}$ in the main paper, we do this for the brevity of the formula). Hope this could help.

**Comment by yaoyuan10475** [comment]
This is very helpful to me. But what form does this regularization take? I currently have two methods. The first is to use ```torch.nn.functional.normalize(L_task, p=2,dim=1)```. The second is $\frac{v - v_{min}}{v_{max} - v_{min}}$。Which form should be used? Or some other method.

**Comment by yaoyuan10475** [comment]
The problem has been solved, thank you very much for your help. @xuanlongORZ @giannifranchi
